cmd/cagent/winui.go

```diff
@@ -285,51 +267,120 @@ func windowsShowSettingsUI(ca *cagent.Cagent, installationMode bool) {
 	ui.MainWindow.Run()
 }
 
-func startService(s *mgr.Service) error {
+func startService(ctx context.Context, s *mgr.Service) error {
 	err := s.Start("is", "manual-started")
 	if err != nil {
+		err = errors.Wrap(err, "could not schedule a service to start")
 		return err
 	}
 
-	return waitServiceState(s, svc.Status{}, svc.Running, time.Second*15)
+	return waitServiceState(ctx, s, svc.Running)
 }
 
-func stopService(s *mgr.Service) error {
+func stopService(ctx context.Context, s *mgr.Service) error {
 	status, err := s.Control(svc.Stop)
 	if err != nil {
+		if strings.Contains(err.Error(), "has not been started") {
+			return nil
+		}
+		err = errors.Wrap(err, "could not schedule a service to stop")
 		return err
 	}
-
-	return waitServiceState(s, status, svc.Stopped, time.Second*15)
+	if status.State == svc.Stopped {
+		return nil
+	}
+	return waitServiceState(ctx, s, svc.Stopped)
 }
 
-func waitServiceState(s *mgr.Service, currentStatus svc.Status, state svc.State, timeout time.Duration) error {
-	stopAt := time.Now().Add(timeout)
-	var err error
-	for currentStatus.State != state {
-		if stopAt.Before(time.Now()) {
-			return fmt.Errorf("timeout waiting for service to stop")
-		}
-		time.Sleep(300 * time.Millisecond)
-		currentStatus, err = s.Query()
-		if err != nil {
-			return fmt.Errorf("could not retrieve service status: %v", err)
+// waitServiceState checks the current state of a service and waits until it will match
+// the expectedState, or a context deadline appearing first.
+func waitServiceState(ctx context.Context, s *mgr.Service, expectedState svc.State) error {
+	for {
+		select {
+		case <-ctx.Done():
+			if ctx.Err() == context.DeadlineExceeded {
+				err := errors.Wrap(ctx.Err(), "timeout waiting for service to stop")
+				return err
+			}
+			return nil
+		default:
+			currentStatus, err := s.Query()
+			if err != nil {
+				err := errors.Wrap(err, "could not retrieve service status")
+				return err
+			}
+			if currentStatus.State == expectedState {
+				return nil
+			}
+			time.Sleep(300 * time.Millisecond)
 		}
 	}
-
 	return nil
 }
 
-func (ui *UI) ShowAdminAlert() {
-	if !checkAdmin() {
-		RunDialog(ui.MainWindow, ui.ErrorIcon, "Error", "Please run as administrator", func() { os.Exit(1) })
+func (ui *UI) CheckPermissions() {
+	if checkPermissions() {
+		return
 	}
+	RunDialog(ui.MainWindow, ui.ErrorIcon,
+		"Error", "Please run this program with administrator priveleges.", func() {
+			os.Exit(1)
+		})
 }
 
-func checkAdmin() bool {
+func checkPermissions() bool {
```

Comment:
well, this func definition looks strange now. It's unclear, what value will be returned. The old name was not good too, but a new one is even harder to understand until you look at the function body.

How about `checkIsAdminUser()` or `checkIsRunningWithElevatedPrivileges()` ?

Reply:
```
checkPermissions bool = true = has permissions
checkPermissions bool = false = doesn't have permissions
```

Running as admin is only a solution to permissions issue, not the prerequisite.
The function itself doesn't check if user is admin, it only checks disk access permissions.

> checkIsRunningWithElevatedPrivileges

This should be a comment for the function, not name :D

Reply:
Researched about `\\\\.\\PHYSICALDRIVE0` notation. This works for me as regular user, without "Run as Administrator", so I assumed this checks disc access only. I will rename to `checkIsRunningWithElevatedPrivileges ` to avoid confusion.